### ISSUE TITLE
[Bug?]: workspaces foreach --include and --exclude do not filter workspaces

### ISSUE THREAD

**Issue by petervannes** [yarnpkg/berry#5505]
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

In a Yarn 3 monorepo i have configured the workspaces `["packages/**","configs/**"],`.
When executing `yarn workspaces list` i do get the expected list of workspaces;

```
➤ YN0000: .
➤ YN0000: configs/babel
➤ YN0000: configs/eslint
➤ YN0000: configs/jest
➤ YN0000: packages/cancel
➤ YN0000: packages/save
```

I want to run a script for one or more specific workspaces using a specific name or using globs.
In the root `package.json` i have several scripts which use `yarn workspaces foreach` with the include argument to run only the script for workspace 'configs/list', none of these seem to work

```
   "build:configs1": "yarn workspaces foreach --verbose  --include configs/list run build",
    "build:configs2": "yarn workspaces foreach --verbose  --include [configs/**] run build",
    "build:configs3": "yarn workspaces foreach --verbose  --include [configs/list] run build",
    "build:configs4": "yarn workspaces foreach --verbose  --include 'configs/list' run build",
    "build:configs5": "yarn workspaces foreach --verbose  --include ['configs/list'] run build",
    "build:configs6": "yarn workspaces foreach --verbose  --include ['configs\/list'] run build",
    "build:configs7": "yarn workspaces foreach --verbose  --include \"configs/list\" run build",
    "build:all": "yarn workspaces foreach --parallel --verbose  run build"
```

configs1 completes without processing any workspaces
configs2 returns 'No matches found: "[configs/**]"'
configs3 returns  'No matches found: "[configs/list]"
configs4 completes without processing any workspaces
configs5 completes without processing any workspaces
configs6 completes without processing any workspaces
configs7 completes without processing any workspaces

### To reproduce

git clone https://github.com/petervannes/yarn3monorepo
yarn install
yarn build:configs1

### Environment

```shell
System:
    OS: macOS 13.5
    CPU: (10) x64 Apple M1 Pro
  Binaries:
    Node: 14.19.1 - /private/var/folders/2t/fpdrn99x4ts52_69rv9_r4yc0000gn/T/xfs-475faa59/node
    Yarn: 3.6.0 - /private/var/folders/2t/fpdrn99x4ts52_69rv9_r4yc0000gn/T/xfs-475faa59/yarn
```

### Additional context

I created a repo to demonstrate the issue https://github.com/petervannes/yarn3monorepo

**Comment by aderant-smcd** [comment]
This is really unintuitive - I would expect a command that operates on workspaces like workspaces foreach to use the same names as declared in the root package.json workspaces array, not the names in the package.json.

**Comment by kumavis** [comment]
for your flag you put `--include configs/list` but I don't know what `list` is referring to here or what you expect that to match

I recommend closing this issue

**Comment by clemyan** [comment]
Using paths to reference workspaces has been available since 4.0.0-rc.41 via #5308
